packages/waku/src/router/create-pages.tsx

```diff
@@ -457,3 +458,331 @@ export function createPages<
     >;
   };
 }
+
+export const new_createPages = <
+  AllPages extends (AnyPage | ReturnType<CreateLayout>)[],
+>(
+  fn: (fns: {
+    createPage: CreatePage;
+    createLayout: CreateLayout;
+    createRoot: CreateRoot;
+  }) => Promise<AllPages>,
+) => {
+  let configured = false;
+
+  const routeMap = new Map<string, Record<string, ReactNode>>();
+
+  // TODO I think there's room for improvement to refactor these structures
+  const fixedPathSet = new Set<[string, PathSpec]>();
+  const dynamicPagePathMap = new Map<
+    string,
+    [PathSpec, FunctionComponent<any>]
+  >();
+  const wildcardPagePathMap = new Map<
+    string,
+    [PathSpec, FunctionComponent<any>]
+  >();
+  const dynamicLayoutPathMap = new Map<
+    string,
+    [PathSpec, FunctionComponent<any>]
+  >();
+  const staticComponentMap = new Map<string, FunctionComponent<any>>();
+  let rootItem: RootItem | undefined = undefined;
+  const noSsrSet = new WeakSet<PathSpec>();
+
+  const registerStaticComponent = (
+    id: string,
+    component: FunctionComponent<any>,
+  ) => {
+    if (
+      staticComponentMap.has(id) &&
+      staticComponentMap.get(id) !== component
+    ) {
+      throw new Error(`Duplicated component for: ${id}`);
+    }
+    staticComponentMap.set(id, component);
+  };
+
+  const createPage: CreatePage = (page) => {
+    if (configured) {
+      throw new Error('createPage no longer available');
+    }
+
+    routeMap.set(page.path, {
+      [`route:${page.path}`]: (
+        // layout adding is different at each level
+        // TODO loop over the path spec to add layout slots
+        <Slot id="root">
+          <Slot id={`layout:${page.path}`}>
+            <Slot id={`page:${page.path}`} />
+          </Slot>
+        </Slot>
+      ),
+      [`page:${page.path}`]: page.component({ children: <Children /> }),
```

Comment:
@dai-shi when are the page component's props supposed to be passed into it? typescript errors here as missing the page props, but I don't believe this is where they would actually be passed in, is it?

Reply:
I think we create an anonymous component, maybe? Can you check how the current `unstable_defineRouter` behaves?